File: srv/resources/site-config/models/action.js
{
  "fields": [
    {
      "meta": {
        "invisible": true
      },
      "type": null,
      "groupName": null,
      "name": "parentId"
    },
    {
      "meta": {
        "invisible": true
      },
      "type": null,
      "groupName": null,
      "name": "caseId"
    },
    {
      "meta": {
        "dictionaryName": "ActionNames",
        "invisible": true
      },
      "type": "dictionary",
      "groupName": null,
      "name": "name"
    },
    {
      "meta": null,
      "type": "statictext",
      "groupName": null,
      "name": "description"
    },
    {
      "meta": {
        "label": "Ожидаемое время выполнения",
        "readonly": true
      },
      "type": "datetime",
      "groupName": null,
      "name": "duetime"
    },
    {
      "meta": {
        "label": "Комментарий"
      },
      "type": "textarea",
      "groupName": null,
      "name": "comment"
    },
    {
      "meta": {
        "label": "Отложить на",
        "dictionaryName": "DeferTimes",
        "regexp": "timespan"
      },
      "type": "dictionary",
      "name": "deferBy"
    },
    {
      "meta": {
        "addClass": "redirectOnChange",
        "label": "Результат",
        "dictionaryName": "ActionResults",
        "dictionaryParent": "name"
      },
      "type": "dictionary",
      "groupName": null,
      "name": "result"
    },
    {
      "meta": {
        "invisible": true
      },
      "type": "datetime",
      "groupName": null,
      "name": "ctime"
    },
    {
      "meta": {
        "invisible": true
      },
      "type": "datetime",
      "groupName": null,
      "name": "mtime"
    },
    {
      "meta": {
        "invisible": true
      },
      "type": "datetime",
      "groupName": null,
      "name": "openTime"
    },
    {
      "meta": {
        "invisible": true
      },
      "type": "datetime",
      "groupName": null,
      "name": "closeTime"
    },
    {
      "meta": {
        "label": "Ответственный",
        "invisible": true,
        "dictionaryType": "BoUsersDict"
      },
      "type": "dictionary",
      "groupName": null,
      "name": "assignedTo"
    },
    {
      "meta": {
        "label": "Роль",
        "invisible": true,
        "dictionaryName": "Roles"
      },
      "type": "dictionary",
      "groupName": null,
      "name": "targetGroup"
    },
    {
      "meta": {
        "label": "Приоритет",
        "invisible": true
      },
      "type": null,
      "groupName": null,
      "name": "priority"
    },
    {
      "meta": {
        "label": "Закрыто",
        "invisible": true
      },
      "type": "checkbox",
      "groupName": null,
      "name": "closed"
    }
  ],
  "applications": [],
  "canDelete": true,
  "canUpdate": true,
  "canRead": true,
  "canCreate": true,
  "title": "Действие",
  "name": "action"
}

File: carma-models/src/Carma/Model/Action.hs
module Carma.Model.Action where

import Data.Aeson as A
import Data.Text
import Data.Time.Clock (UTCTime)
import Data.Typeable

import Data.Model
import Data.Model.View

import Carma.Model.Types (HMDiffTime)

import Carma.Model.ActionResult (ActionResult)
import Carma.Model.ActionType (ActionType)
import Carma.Model.Case (Case)
import Carma.Model.DeferTime (label, time)
import Carma.Model.Role (Role)
import Carma.Model.Service (Service)
import Carma.Model.Usermeta (Usermeta)

data Action = Action
  { ident       :: PK Int Action                    "Действие"
  , caseId      :: F (IdentI Case)                  "caseId" "Кейс"
  , serviceId   :: F (Maybe (IdentI Service))       "serviceId" "Услуга"
  , aType       :: F (IdentI ActionType)            "type" "Тип действия"
  , duetime     :: F UTCTime                        "duetime" "Ожидаемое время выполнения"
  , comment     :: F (Maybe Text)                   "comment" "Комментарий"
  , deferBy     :: F (Maybe HMDiffTime)             "deferBy" "Отложить на"
  , result      :: F (Maybe (IdentI ActionResult))  "result" "Результат"
  , ctime       :: F UTCTime                        "ctime" "Время создания"
  , assignTime  :: F (Maybe UTCTime)                "assignTime" "Время назначения"
  , openTime    :: F (Maybe UTCTime)                "openTime" "Время начала работы"
  , closeTime   :: F (Maybe UTCTime)                "closeTime" "Время закрытия"
  , assignedTo  :: F (Maybe (IdentI Usermeta))      "assignedTo" "Ответственный"
  , targetGroup :: F (IdentI Role)                  "targetGroup" "Роль"
  , parent      :: F (Maybe (IdentI Action))        "parent" "Предыдущее действие"
  } deriving Typeable

instance Model Action where
  type TableName Action = "actiontbl"
  modelInfo = mkModelInfo Action ident
  modelView = \case
    "" -> Just $ modifyView defaultView $
          [ hiddenIdent caseId
          , hiddenIdent serviceId
          , hiddenIdent parent
          , readonly duetime
          , deferBy `completeWith` time
          , setMeta "dictionaryLabel"
            (A.String $ fieldName label) deferBy
          , infoText "defertime" deferBy
          , setMeta "addClass" "redirectOnChange" result
          , setMeta "dictionaryType" "BoUsersDict" assignedTo
          , setMeta "dictionaryType" "ResultsDict" result
          , setMeta "dictionaryParent"
            (A.String $ fieldName aType) result
          , setType "textarea" comment
          , invisible aType
          , invisible ctime
          , invisible assignTime
          , invisible openTime
          , invisible closeTime

            -- These two are overrided by client code on #supervisor
            -- screen
          , invisible assignedTo
          , invisible targetGroup

          , dict targetGroup $ (dictOpt "backofficeRoles")
            { dictType = Just "ComputedDict"
            , dictBounded = True
            }
          ]
    _  -> Nothing
